File: Algorithm/code/exp-2-递归与分治策略/2. n个互异元素的全排列问题（P13）/C/main.c
#include <stdio.h>

void Swap(char *a, char *b)
{
    char temp = *a;
    *a = *b;
    *b = temp;
}

/*
 *  产生 list[k:m] 的所有排列 
 */
void Perm(char list[], int k, int m)
{
    // 只剩下一个元素
    if(k == m)
    {       
        for(int i=0; i<=m; i++)
            printf("%c ", list[i]);
        printf("\n");
    }
    else
    {
        for(int i=k; i<=m; i++)
        {
            Swap(&list[k], &list[i]);
            Perm(list, k+1, m);
            Swap(&list[k], &list[i]);
        }
    }    
}

int main()
{
    char list[4] = {'a', 'b', 'e', 'f'};
    Perm(list, 0, 3);

    return 0;
}

File: Algorithm/code/exp-2-递归与分治策略/2. n个互异元素的全排列问题（P13）/Python/main.py
def Perm(list, k, m):
    if k == m:
        for i in range(m+1):
            print(str(list[i]) + ' ', end="")
        print()

    else:
        for i in range(k, m+1):
            list[k], list[i] = list[i], list[k]
            Perm(list, k+1, m)
            list[k], list[i] = list[i], list[k]

list = ['a', 'b', 'e', 'f']
Perm(list, 0, 3)